torch/csrc/api/include/torch/detail/member_ref.h

```diff
@@ -0,0 +1,68 @@
+#include <torch/parameter_list.h>
+
+#include <torch/csrc/autograd/variable.h>
+
+#include <functional>
+#include <type_traits>
+
+namespace torch { namespace detail {
+
+/// A class that stores const-correct getters to a member variable, accessible
+/// through a pointer to an object.
+template <typename T>
+class MemberRef {
+ public:
+  // TODO: Replace with (std/boost/our)::any.
+  using This = void*;
+  using ConstThis = const void*;
+  using Getter = std::function<T&(This)>;
+  using ConstGetter = std::function<const T&(ConstThis)>;
+
+  template <typename Class>
+  /* implicit */ MemberRef(T Class::*member)
+      : getter_([member](This self) -> T& {
+          return static_cast<Class*>(self)->*member;
+        }),
+        const_getter_([member](ConstThis self) -> const T& {
+          return static_cast<const Class*>(self)->*member;
+        }) {}
```

Comment:
I'm slightly concerned about the cost of keeping those lambdas around and calling them so much. Wouldn't it be simpler to keep an `offsetof` value, and then adjust the pointer value of `self` in the getters?

Reply:
Yeah, you're right. That approach is even more dangerous, but it's probably worth it and will simplify the code. I'll try it out

Reply:
Actually, I think this wouldn't work polymorphically would it? Right now, when `this` points to a derived class, but we cast it to the base class inside, it's ok. With just offsets, this would go wrong

Reply:
Well your thing doesn't work polymorphically either. You need to case the pointer to a `Class` that's fixed at compile time, so nothing really changes.